packages/cursorless-engine/src/util/uniqWithHash.ts

```diff
@@ -0,0 +1,74 @@
+import { uniqWith, isEqual } from "lodash";
+
+/**
+ * Like lodash.uniqWith, but uses a hash function to (mostly) avoid quadratic runtime.
+ * @param array The array to uniq
+ * @param fn The equality function
+ * @param hash The hash function. It must be a valid hash function, insofar as it must return the same value for equal items. A hash function that returns a constant value is equivalent to lodash.uniqWith.
+ * @returns The uniq array
+ */
+export function uniqWithHash<T>(
+  array: T[],
+  fn: (a: T, b: T) => boolean,
+  hash: (t: T) => string,
+): T[] {
+  // Handle the common, tiny cases without allocating anything extra.
+  if (array.length < 2) {
+    return [...array];
+  }
+  if (array.length === 2) {
+    if (fn(array[0]!, array[1]!)) {
+      return [array[0]!];
+    }
+    return [...array];
+  }
+  // First, split up the array using the hash function.
+  // This keeps the sets of items passed to uniqWith small,
+  // so that the quadratic runtime of uniqWith less of a problem.
+  // Keep track of which sets have multiple items, so that we can uniq them.
```

Comment:
I think this line should be a JSDoc so we get hover hints on `needsUniq`
```suggestion
  /** Keep track of which sets have multiple items, so that we can uniq them. */
```

Reply:
done